numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,662 @@
+                    if len(key) == 1:
+                        return key
+                    raise IndexError("Boolean array indices combined with other indices are not allowed in the array API namespace")
+
+            if shape is None:
+                return key
+            n_ellipsis = key.count(...)
+            if n_ellipsis > 1:
+                return key
+            ellipsis_i = key.index(...) if n_ellipsis else len(key)
+
+            for idx, size in list(zip(key[:ellipsis_i], shape)) + list(zip(key[:ellipsis_i:-1], shape[:ellipsis_i:-1])):
+                ndarray._validate_index(idx, (size,))
+            return key
+        elif isinstance(key, bool):
+            return key
+        elif isinstance(key, ndarray):
+            if key.dtype in _integer_dtypes:
+                if key.shape != ():
+                    raise IndexError("Integer array indices with shape != () are not allowed in the array API namespace")
+            return key._array
+        elif key is Ellipsis:
+            return key
+        elif key is None:
+            raise IndexError("newaxis indices are not allowed in the array API namespace")
+        try:
+            return operator.index(key)
+        except TypeError:
+            # Note: This also omits boolean arrays that are not already in
+            # ndarray() form, like a list of booleans.
+            raise IndexError("Only integers, slices (`:`), ellipsis (`...`), and boolean arrays are valid indices in the array API namespace")
+
+    def __getitem__(self: array, key: Union[int, slice, Tuple[Union[int, slice], ...], array], /) -> array:
+        """
+        Performs the operation __getitem__.
+        """
+        # Note: Only indices required by the spec are allowed. See the
+        # docstring of _validate_index
+        key = self._validate_index(key, self.shape)
+        res = self._array.__getitem__(key)
+        return self.__class__._new(res)
+
+    def __gt__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __gt__.
+        """
+        res = self._array.__gt__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __int__(self: array, /) -> int:
+        """
+        Performs the operation __int__.
+        """
+        # Note: This is an error here.
+        if self._array.shape != ():
+            raise TypeError("bool is only allowed on arrays with shape ()")
+        res = self._array.__int__()
+        return res
+
+    def __invert__(self: array, /) -> array:
+        """
+        Performs the operation __invert__.
+        """
+        res = self._array.__invert__()
+        return self.__class__._new(res)
+
+    def __le__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __le__.
+        """
+        res = self._array.__le__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __len__(self, /):
```

Comment:
Comment in the standard:
_TODO: need to more carefully consider this in order to accommodate, e.g., graph tensors where a shape may be dynamic. Furthermore, not clear whether this should be implemented, as, e.g., NumPy's behavior of returning the size of the first dimension is not necessarily intuitive, as opposed to, say, the total number of elements._

I think removing it may make sense. Writing `len(x)` is never a good idea I believe; `x.shape[0]` or `x.size` is much clearer.

Reply:
Also worth pointing out that Python restricts `__len__` so it has to return an `int`. Returning a dynamic shape object from `__len__` is impossible.

Reply:
Well sympy.Matrix is weird. It is fixed at dimension 2, meaning you can't get a 1-D subarray or anything. It's a mathematical matrix which is always n x m. As a result it has weird indexing, which doesn't match NumPy. Personally I would prefer if it tried to match NumPy indexing a little better. 

Honestly omitting `__len__` and `__iter__` from the spec is probably the best. You're right that iterating all elements and having len == size is nice, but that breaks `list(x.__iter__()) == [x[i] for i in range(len(x))]`.

Anyway, this is all more of a spec question, so it really should be discussed there.